transport.go

```diff
@@ -24,19 +40,27 @@ type Conn interface {
 // but many more can be implemented, sctp, audio signals, sneakernet, UDT, a
 // network of drones carrying usb flash drives, and so on.
 type Transport interface {
-	Dialer(laddr ma.Multiaddr, opts ...DialOpt) (Dialer, error)
+	// Dial dials a remote peer. It should try to reuse local listener
+	// addresses if possible but it may choose not to.
+	Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (Conn, error)
+
+	// CanDial returns true if this transport knows how to dial the given
+	// multiaddr.
+	//
+	// Returning true does not guarantee that dialing this multiaddr will
+	// succeed. This function should *only* be used to preemptively filter
+	// out addresses that we can't dial.
+	CanDial(addr ma.Multiaddr) bool
+
+	// Listen listens on the passed multiaddr.
 	Listen(laddr ma.Multiaddr) (Listener, error)
-	Matches(ma.Multiaddr) bool
-}
 
-// Dialer is an abstraction that is normally filled by an object containing
-// information/options around how to perform the dial. An example would be
-// setting TCP dial timeout for all dials made, or setting the local address
-// that we dial out from.
-type Dialer interface {
-	Dial(raddr ma.Multiaddr) (Conn, error)
-	DialContext(ctx context.Context, raddr ma.Multiaddr) (Conn, error)
-	Matches(ma.Multiaddr) bool
+	// Protocol returns the set of protocols handled by this transport.
+	Protocols() []int
+
+	// Proxy returns true if this transport proxies.
+	// TODO: Make this a part of the go-multiaddr protocol instead?
```

Comment:
I think it's fine to leave this here for now but we may want to move it later. Proxy transports are transports like `p2p-circuit` that handle part of the multiaddr and forward the rest to some other system.